lib/Backend/Lower.cpp

```diff
@@ -23001,6 +22972,242 @@ Lowerer::LoadIndexFromLikelyFloat(
+{
+    //   CMP [objectOpnd + offset(objectArray)], nullptr
+    //   JEQ $noObjectArrayLabel
+    //   TEST[objectOpnd + offset(objectArray)], ObjectArrayFlagsTag        (used as flags)
+    //   JEQ $noObjectArrayLabel
+    //   MOV typeHandlerOpnd, [typeOpnd + offset(typeHandler)]
+    //   CMP typeHandler->OffsetOfInlineSlots, Js::DynamicTypeHandler::GetOffsetOfObjectHeaderInlineSlots()
+    //   JNE $hasObjectArrayLabel
+    // $$noObjectArrayLabel:                                                (fall thru)
+
+    Func * func = this->m_func;
+    IR::LabelInstr * noObjectArrayLabel = IR::LabelInstr::New(Js::OpCode::Label, func);
+    IR::IndirOpnd * objectArrayOpnd = IR::IndirOpnd::New(objectOpnd, Js::DynamicObject::GetOffsetOfObjectArray(), TyMachPtr, func);
+    InsertCompareBranch(objectArrayOpnd, IR::AddrOpnd::NewNull(func), Js::OpCode::BrEq_A, noObjectArrayLabel, insertBeforeInstr);
+    InsertTestBranch(objectArrayOpnd, IR::IntConstOpnd::New((uint32)Js::DynamicObjectFlags::ObjectArrayFlagsTag, TyUint8, func), 
+        Js::OpCode::BrNeq_A, noObjectArrayLabel, insertBeforeInstr);
+
+    IR::RegOpnd * typeHandlerOpnd = IR::RegOpnd::New(TyMachPtr, func);
+    InsertMove(typeHandlerOpnd, IR::IndirOpnd::New(typeOpnd, Js::DynamicType::GetOffsetOfTypeHandler(), TyMachPtr, func), insertBeforeInstr);
+    InsertCompareBranch(IR::IndirOpnd::New(typeHandlerOpnd, Js::DynamicTypeHandler::GetOffsetOfOffsetOfInlineSlots(), TyUint16, func),
+        IR::IntConstOpnd::New(Js::DynamicTypeHandler::GetOffsetOfObjectHeaderInlineSlots(), TyUint16, func), 
+        Js::OpCode::BrNeq_A, hasObjectArrayLabel, insertBeforeInstr);
+
+    insertBeforeInstr->InsertBefore(noObjectArrayLabel);
+}
+
+void
+Lowerer::GenerateInitForInEnumeratorFastPath(IR::Instr * instr, Js::ForInCache * forInCache)
+{
+    Func * func = this->m_func;
+
+    IR::LabelInstr * helperLabel = IR::LabelInstr::New(Js::OpCode::Label, func, true);
+    IR::RegOpnd * objectOpnd = instr->GetSrc1()->AsRegOpnd();
+
+    // Tagged check and object check
+    m_lowererMD.GenerateObjectTest(objectOpnd, instr, helperLabel);
+    m_lowererMD.GenerateIsDynamicObject(objectOpnd, instr, helperLabel);
+
+    // Type check with cache
+    //
+    //   MOV typeOpnd, [objectOpnd + offset(type)]
+    //   CMP [&forInCache->type], typeOpnd
+    //   JNE $helper
+
+    IR::RegOpnd * typeOpnd = IR::RegOpnd::New(TyMachPtr, func);
+    InsertMove(typeOpnd, IR::IndirOpnd::New(objectOpnd, Js::DynamicObject::GetOffsetOfType(), TyMachPtr, func), instr);
+    InsertCompareBranch(IR::MemRefOpnd::New(&forInCache->type, TyMachPtr, func, IR::AddrOpndKindForInCacheType), typeOpnd, Js::OpCode::BrNeq_A, helperLabel, instr);
+    
+    // Check forInCacheData->EnumNonEnumerable == false  
+    //
+    //   MOV forInCacheDataOpnd, [&forInCache->data]
+    //   CMP forInCacheDataOpnd->enumNonEnumerable, 0
+    //   JNE $helper
+
+    IR::RegOpnd * forInCacheDataOpnd = IR::RegOpnd::New(TyMachPtr, func);
```

Comment:
No..  FI the forInCache->type manages the object's type, that means the type can't be null, therefore, the data can't be null either.